src/pudl/extract/excel.py

```diff
@@ -354,3 +363,103 @@ def excel_filename(self, page, **partition):
+    and values are DataFrame of extracted data.
+    Where a page name appears more than once in the dictionaries, all
+    DataFrames with that page name are merged together, and the result
+    returned as a single dict mapping unique page names to merged
+    DataFrames.
+    """
+    merged = defaultdict(list)
+    for dfs in paged_dfs:
+        for page in dfs:
+            merged[page].append(dfs[page])
+
+    for page in merged:
+        merged[page] = pd.concat(merged[page]).reset_index(drop=True)
+
+    return merged
+
+
+def year_loader_factory(
+    extractor_cls: type[GenericExtractor], name: str
+) -> OpDefinition:
+    """Return a dagster op to load a single year using a given extractor class.
+
+    Args:
+        extractor_cls: Class of type GenericExtractor used to extract the data.
+        name(str): Name of dataset (e.g. "eia860").
+    """
+
+    def load_single_year(context, year: int) -> dict[str, pd.DataFrame]:
+        """Load a single year of EIA data from file.
+
+        Args:
+            context:
+                context: dagster keyword that provides access to resources and config.
+            year:
+                Year to load.
+
+        Returns:
+            Loaded data in a dataframe.
+        """
+        ds = context.resources.datastore
+        return extractor_cls(ds).extract(year=[year])
+
+    return op(
+        required_resource_keys={"datastore", "dataset_settings"},
+        name=f"load_single_{name}_year",
+    )(load_single_year)
+
+
+def years_from_settings_factory(name: str) -> OpDefinition:
+    """Return a dagster op to get EIA years from settings.
+
+    Args:
+        name (str): Name of dataset (e.g. "eia860")
+    """
+
+    def years_from_settings(context):
+        """Return set of years for EIA data in settings.
+
+        These will be used to kick off worker processes to load each year of data in
+        parallel.
+        """
+        eia_settings = context.resources.dataset_settings.eia
+        for year in getattr(eia_settings, name).years:
+            yield DynamicOutput(year, mapping_key=str(year))
+
+    return op(
+        out=DynamicOut(),
+        required_resource_keys={"dataset_settings"},
+        name=f"{name}_years_from_settings",
+    )(years_from_settings)
+
+
+def raw_df_factory(
+    extractor_cls: type[GenericExtractor], name: str
+) -> AssetsDefinition:
+    """Return a graph asset to load a set of raw DataFrames from Excel files.
+
+    Args:
+        extractor_cls: Class of type GenericExtractor used to extract the data.
+        name(str): Name of dataset (e.g. "eia860").
+    """
+    year_loader = year_loader_factory(extractor_cls, name)
+    years_from_settings = years_from_settings_factory(name)
+
+    def raw_dfs() -> dict[str, pd.DataFrame]:
+        """All loaded EIA dataframes."""
+        years = years_from_settings()
+        dfs = years.map(lambda year: year_loader(year))
+        return merge_dfs_by_page(dfs.collect())
```

Comment:
Is there a reason it's better to use `map()` and `collect()` here rather than something like a list comprehension? I don't think I really understand how these work and find it hard to read.

Reply:
After some poking: these are actually calls to the Dagster `map` and `collect` orchestration methods - we can't iterate over dynamic outputs in a list so list comprehension wouldn't work here. https://docs.dagster.io/_apidocs/dynamic